src/runtime_src/core/common/scope_guard.h

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef core_common_scope_guard_h
+#define core_common_scope_guard_h
+
+#include <utility>
+#include <type_traits>
+
+namespace xrt_core { 
+
+/**
+ * class scope_guard - RAII for fundamental types that need to be 
+ * terminated / cleaned-up at scope exit
+ */
+template <typename ValueType, typename ExitFunction>
+class scope_guard
+{
+  static_assert(std::is_fundamental<ValueType>::value,"Invalid ValueType");
+  ValueType value;
+  ExitFunction exfcn;
```

Comment:
Since `ExitFunction` can be any callable including a lambda, I am unsure we need a `value` at all

Reply:
We need a value for sure. I realize the lambda and that was my first approach.  However, the guard is there to manage lifetime of the 'value' while providing access to the 'value' for clients who need it.  Before the guard, it was

> int value = open(...);
> DoSomething(value);
> close(value);

with the guard it is

> auto v = open_as_a_guard(...);
> DoSomething(v.get());

C++ is in dire need of a generic scope guard.   I do not necessarily like what I cooked up, but I also do not like the unguarded approach.